File: dbrepo-metadata-service/services/src/main/java/at/tuwien/service/OntologyService.java
package at.tuwien.service;

import at.tuwien.api.semantics.OntologyCreateDto;
import at.tuwien.api.semantics.OntologyModifyDto;
import at.tuwien.entities.semantics.Ontology;
import at.tuwien.exception.OntologyNotFoundException;

import java.security.Principal;
import java.util.List;
import java.util.UUID;

public interface OntologyService {

    /**
     * Finds all ontologies in the metadata database.
     *
     * @return The list of ontologies.
     */
    List<Ontology> findAll();

    /**
     * Finds all processable ontologies.
     *
     * @return The list of ontologies.
     */
    List<Ontology> findAllProcessable();

    /**
     * Finds an ontology in the metadata database with given id.
     *
     * @param ontologyId The ontology id.
     * @return The ontology, if successful.
     * @throws OntologyNotFoundException The ontology was not found in the metadata database.
     */
    Ontology find(UUID ontologyId) throws OntologyNotFoundException;

    Ontology find(String entityUri) throws OntologyNotFoundException;

    /**
     * Registers an ontology in the metadata database.
     *
     * @param data      The ontology data.
     * @param principal The user principal.
     * @return The created ontology, if successful.
     */
    Ontology create(OntologyCreateDto data, Principal principal);

    /**
     * Updates an ontology in the metadata database with given id.
     *
     * @param ontology The ontology.
     * @param data     The ontology data.
     * @return The updated ontology, if successful.
     */
    Ontology update(Ontology ontology, OntologyModifyDto data);

    /**
     * Unregisters an ontology in the metadata database with given id.
     *
     * @param ontology The ontology.
     */
    void delete(Ontology ontology);
}

File: dbrepo-ui/composables/ontology-service.ts
export const useOntologyService = (): any => {
  async function findAll(): Promise<OntologyDto[]> {
    const axios = useAxiosInstance()
    console.debug('find ontologies')
    return new Promise<OntologyDto[]>((resolve, reject) => {
      axios.get<OntologyDto[]>('/api/semantic/ontology')
        .then((response) => {
          console.info(`Found ${response.data.length} ontology(s)`)
          resolve(response.data)
        })
        .catch((error) => {
          console.error('Failed to find ontologies', error)
          reject(error)
        })
    })
  }

  async function findOne(id: number): Promise<OntologyDto> {
    const axios = useAxiosInstance()
    console.debug('find ontology for id', id)
    return new Promise<OntologyDto>((resolve, reject) => {
      axios.get<OntologyDto>(`/api/semantic/ontology/${id}`)
        .then((response) => {
          console.info('Found ontology for id', id)
          resolve(response.data)
        })
        .catch((error) => {
          console.error('Failed to find ontology', error)
          reject(error)
        })
    })
  }

  async function create(data: OntologyCreateDto): Promise<OntologyDto> {
    const axios = useAxiosInstance()
    console.debug('create ontology')
    return new Promise<OntologyDto>((resolve, reject) => {
      axios.post<OntologyDto>('/api/semantic/ontology', data)
        .then((response) => {
          console.info('Created ontology with id', response.data.id)
          resolve(response.data)
        })
        .catch((error) => {
          console.error('Failed to create ontology', error)
          reject(error)
        })
    })
  }

  async function update(id: number, data: OntologyModifyDto): Promise<OntologyDto> {
    const axios = useAxiosInstance()
    console.debug('update ontology with id', id)
    return new Promise<OntologyDto>((resolve, reject) => {
      axios.put<OntologyDto>(`/api/semantic/ontology/${id}`, data)
        .then((response) => {
          console.info('Updated ontology with id', id)
          resolve(response.data)
        })
        .catch((error) => {
          console.error('Failed to update ontology', error)
          reject(error)
        })
    })
  }

  async function remove(id: number): Promise<void> {
    const axios = useAxiosInstance()
    console.debug('delete ontology with id', id)
    return new Promise<void>((resolve, reject) => {
      axios.delete<void>(`/api/semantic/ontology/${id}`)
        .then(() => {
          console.info('Deleted ontology with id', id)
          resolve()
        })
        .catch((error) => {
          console.error('Failed to delete ontology', error)
          reject(error)
        })
    })
  }

  return {findAll, findOne, create, update, remove}
}
